library/std/src/sys/pal/unix/kernel_copy.rs

```diff
@@ -619,7 +610,43 @@ pub(super) fn copy_regular_files(reader: RawFd, writer: RawFd, max_len: u64) ->
                 return match err.raw_os_error() {
                     // when file offset + max_length > u64::MAX
                     Some(EOVERFLOW) => CopyResult::Fallback(written),
-                    Some(ENOSYS | EXDEV | EINVAL | EPERM | EOPNOTSUPP | EBADF) if written == 0 => {
+                    Some(raw_os_error @ (ENOSYS | EXDEV | EINVAL | EPERM | EOPNOTSUPP | EBADF))
+                        if written == 0 =>
+                    {
+                        if !have_probed {
+                            let available = match raw_os_error {
+                                EPERM => {
+                                    // EPERM can indicate seccomp filters or an
+                                    // immutable file. To distinguish these
+                                    // cases we probe with invalid file
+                                    // descriptors which should result in EBADF
+                                    // if the syscall is supported and EPERM or
+                                    // ENOSYS if it's not available.
+                                    match unsafe {
+                                        cvt(copy_file_range(
+                                            INVALID_FD,
+                                            ptr::null_mut(),
+                                            INVALID_FD,
+                                            ptr::null_mut(),
+                                            1,
+                                            0,
+                                        ))
+                                        .map_err(|e| e.raw_os_error())
+                                    } {
+                                        Err(Some(EPERM | ENOSYS)) => UNAVAILABLE,
+                                        Err(Some(EBADF)) => AVAILABLE,
+                                        Ok(_) => panic!("unexpected copy_file_range probe success"),
+                                        // Treat other errors as the syscall
+                                        // being unavailable.
+                                        Err(_) => UNAVAILABLE,
+                                    }
+                                }
+                                ENOSYS => UNAVAILABLE,
```

Comment:
Is it possible for a fuse fs to return `ENOSYS` to the user even if the kernel has support for the syscall?

Reply:
I assume so, but what does the author of the FUSE FS expect to happen when they do this? I think we can pretend it doesn't happen.

Reply:
We don't have to be robust against the kernel misbehaving, but IMO we should be robust against other processes on the system misbehaving.

Reply:
I'd like to challenge that. Where does it end? You can modify the results of arbitrary syscalls using ptrace. We can't defend against that. Why is FUSE special?

Reply:
It's not, our program might be ptraced.

> The probe-first approach is immune to any such file-dependent questions because it uses invalid FDs so it never touches any filesystem, it only checks kernel support.

This can obviously be listed as part of the requirements, then it can also be done with the current approach.

Reply:
> You can modify the results of arbitrary syscalls using ptrace.

On Ubuntu (and any other distro which configures yama lsm for this) you can only ptrace a program you spawned yourself.

Reply:
Other code in std behaves the same way:

https://github.com/rust-lang/rust/blob/d255c6a57c393db6221b1ff700daea478436f1cd/library/std/src/sys/pal/unix/fs.rs#L199-L202

I think this is not a problem of this particular PR.